### ISSUE TITLE
2FA / MFA

### ISSUE THREAD

**Issue by KidkArolis** [expo/expo#160]
Tried running `exp build:ios`, but credential verification failed since I have 2FA enabled on my Apple account.

**Comment by brentvatne** [comment]
cc @skevy @dikaiosune

**Comment by brentvatne** [comment]
hi @KidkArolis! a couple of us discussed this on twitter yesterday: https://twitter.com/Jackhanford/status/860909996933996544

tl;dr: 2fa not supported currently by ios builder ( 😢 ) but on roadmap

**Comment by 4815162342** [comment]
Versions of OS X that were released before TFA was supported (e.g. 10.10) would have you input the TFA code after the password in the same box. For example, if your password was reactRox and the TFA code was 196 601, then you would input reactRox196601 into the password field.

Has anyone tried this method? It's still definitely a hack, but will prevent you from disabling TFA.

**Comment by terribleben** [comment]
Thanks for the heads up @jonandrewdavis, we weren't aware that Apple made that change. We have somebody working on this in the upcoming week.

**Comment by dikaiosune** [comment]
I'm currently working on this right now, going to close this in favor of the Canny request: https://expo.canny.io/feature-requests/p/allow-to-build-ios-binaries-when-using-apples-2fa. Thanks!

**Comment by 4815162342** [comment]
Thanks Adam!

**Comment by PanRada** [comment]
hi any update ?

**Comment by gregblass** [comment]
Just an update in the mean time as to how I was able to get around this: If you create an apple ID through the website (and not through iOS or Mac OS), 2FA will be off by default. Then if you call apple developer support and explain to them the situation, I was able to have them issue me a refund on the $100 I paid for my developer account associated with my Apple ID that can’t turn off 2FA. Then you can apply again with the account that has 2FA turned off and be good to go.

**Comment by jonandrewdavis** [comment]
Unfortunately, getting a refund and applying again is not a suitable work-around for many.

Looking forward to a potential fix for this issue 👍.

**Comment by gregblass** [comment]
I agree. I had no choice though. That was my last hope. And it took hours of my time pleading between different apple departments.

**Comment by ryantallmadge** [comment]
Can you just eject the project and do it the `old fashion` way?

**Comment by QuanTumli** [comment]
Hi, any updates for 2FA? Whats the progress of the work @dikaiosune?

**Comment by davidthornton** [comment]
@ryantallmadge -- to be explicit as to what the "old fashioned" way is (I had to have this explained to me lol) -- it's `expo detach` or `yarn eject`. I hope this helps someone else who needs to build binaries or do CI!

**Comment by lucasapereira** [comment]
I really don´t want to detach. Any update on this? My app is ready for production.

**Comment by c316** [comment]
@lucasapereira 

I also don't want to detach so for now I've been following these instructions (which I wrote up when I went through the process). 

This way I don't have to detach. It is a pain, no doubt, but it prevents me from having to detach or keep two code bases. 

https://gist.github.com/c316/a877a240e95e64a926a63f460daf92e0

TL;DR
I make a copy of the folder, then fix all the annoying problems that arise from using CRNA and deploy.

**Comment by lifeiscontent** [comment]
@brentvatne is this currently being worked on? If not, I'd be interested in looking into it, I just have no clue where the authentication code is, so if you can point me to it, I'd appreciate it! 👍

**Comment by soukupl** [comment]
You can create a new Apple ID from web without 2FA and add that account to your itunes connect as developer with rights to submit app updates... This will work as "easy" workaround...

**Comment by lifeiscontent** [comment]
@soukupl it should be relatively easy to support app passwords within expo.

**Comment by lifeiscontent** [comment]
@soukupl do you have an example of the process you're speaking of?

**Comment by lifeiscontent** [comment]
@soukupl does this process work for individual developer accounts?

**Comment by soukupl** [comment]
@lifeiscontent yes, it can only be available to company itunes connect accounts... App Password should also work... Not sure what is needed on the Expo side to enable it... Maybe you just generate app password and pass it to expo when logging it... It may be supported right out of the box by Apple login...

**Comment by lifeiscontent** [comment]
@soukupl app passwords don’t work out of the box currently but it’s such low hanging fruit I’d be willing to put some time into it if someone can point me to the right place

**Comment by dikaiosune** [comment]
@lifeiscontent app-specific passwords only work for uploading to iTC with fastlane (the tools we use to power the build service). :(

That said, we have made progress on supporting 2FA authentication, hopefully we'll have an announcement in a few weeks.

**Comment by lifeiscontent** [comment]
@dikaiosune thanks for the update.

**Comment by Raveler** [comment]
Wow, this is a depressing hiccup right before I'm ready to submit my apps to the stores. Android is ready to go but none of the above solutions feel like a feasible option to get the app on iOS right now :(.

**Comment by PanRada** [comment]
@Raveler yes it true is big frustated if you push in your company(big company) RN with Expo and you have apps to push to AppStore but u can not :( hope some one will finish that or we can make some community support. Still wondering where is problem and if some one can point code to fix

**Comment by ide** [comment]
We're working on 2FA and it is taking awhile partly because connecting to iTunes Connect via Fastlane is particularly difficult on Windows. You can detach and build the app yourself if you need 2FA and need to submit to the App Store soon -- that way your launch schedule isn't dependent on 2FA support built into Expo.

**Comment by dikaiosune** [comment]
You can also build a standalone app yourself (without detaching) if you follow instructions in the README of this repository.

**Comment by Raveler** [comment]
That's interesting! I will look into that.

**Comment by lifeiscontent** [comment]
@ide any updates here?

**Comment by ide** [comment]
No updates to share.

**Comment by dflourusso** [comment]
Apple now only supports 2FA, some news?

**Comment by scally** [comment]
@dikaiosune I've tried using non-detached standalone route, but either I've screwed up somehow or the instructions are outdated, as every time I try running the `build` command for ios, it fails with a nonzero exit code -- yet running the bash command it generates manually succeeds for me. 

It's been pretty hard to troubleshoot causation so far. 

Is anyone else able to get the workaround, a non-detached standalone iOS build, working?

```
tools-public 🍀  gulp ios-shell-app --action build --type archive --configuration Release
[00:08:51] Using gulpfile ~/projects/expo/tools-public/gulpfile.js
[00:08:51] Starting 'ios-shell-app'...
Building shell app under ../shellAppBase-archive:

xcodebuild -workspace Exponent.xcworkspace -scheme Exponent -sdk iphoneos -destination generic/platform=iOS -configuration Release archive -derivedDataPath ../shellAppBase-archive -archivePath ../shellAppBase-archive/Exponent.xcarchive CODE_SIGN_IDENTITY="" CODE_SIGNING_REQUIRED=NO | xcpretty

xcodebuild is running. Logging errors only. To see full output, use --verbose 1...
** ARCHIVE FAILED **


The following build commands failed:
	PhaseScriptExecution Generate\ Dynamic\ Macros /Users/sean/projects/expo/shellAppBase-archive/Build/Intermediates.noindex/ArchiveIntermediates/Exponent/IntermediateBuildFilesPath/Exponent.build/Release-iphoneos/Exponent.build/Script-78CEE2F21ACDF25E0095B124.sh
(1 failure)

Finished building, copying artifact to ../shellAppBase-builds/archive/Release...
cp: ../shellAppBase-archive/Exponent.xcarchive: No such file or directory
[00:17:28] 'ios-shell-app' errored after 8.62 min
[00:17:28] Error: Process exited with non-zero code: 1
    at ChildProcess.child.on (/Users/sean/projects/xdl/node_modules/@expo/spawn-async/build/spawnAsync.js:39:21)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at maybeClose (internal/child_process.js:927:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
```

**Comment by MoOx** [comment]
It's weird to close this issue (which is linked from the documentation) :) Hope you guys will succeed into making that possible. Thanks for your work!

**Comment by bcostabatista** [comment]
Any news?

**Comment by ide** [comment]
We are working on it and will announce when it is ready.

**Comment by FelipeBohnertPaetzold** [comment]
Are you forecast to complete this release?

**Comment by ericvan76** [comment]
Do you guys have ETA on this? or can you guys keep this issue open? I appreciate what you're doing, but I think most of developers here is waiting for an answer to decide whether I should wait or just detach from expo to work around.

**Comment by dikaiosune** [comment]
Hi, as I mentioned in the [comment closing this issue}(https://github.com/expo/expo/issues/160#issuecomment-311761083) we're tracking this on the feature request board.

I don't have an exact date on when we'll be able to release this, but we're still actively working on it. Hopefully before the end of the year it'll be available, but it is tricky and takes time to get right, so we can't make any promises.

**Comment by scally** [comment]
Is there an open public PR or some other way we can contribute?

**Comment by dikaiosune** [comment]
@scally not at the moment, but we expect to land this very soon.

**Comment by dikaiosune** [comment]
Hi everyone! Thanks to lots of hard work from @fxfactorial we have preliminary support for Apple 2FA when running standalone builds! We really appreciate everyone's patience while we worked to get this right, we all know that it's been a huge pain point for the whole community.

Along with supporting 2FA, there are some extra benefits of doing this:

* all authentication with Apple is done on your machine, we no longer store passwords on our servers when using the new auth flow (we still store certificates encrypted so you don't need to send them every time you run a build, we might make that optional in the future though)
* error messages related to authentication or fetching certificates should be much more transparent
* it's no longer necessary to copy/paste your team ID from Apple's website which should prevent lots of failed builds

Because this is a big change to how we handle credentials in the standalone flow, it's opt-in for now (as of `exp` 46.0.5). Once we're sure this works well for everyone (including Windows users) we'll go ahead and make it the default for all builds. In the meantime, you can enable it with the `--local-auth` flag for your build:

```
# make sure you have the latest exp
$ exp --version
46.0.5
$ exp build:ios --local-auth
```

Please open new issues on this repository if you run into problems while using the new auth flow!

**Comment by ericvan76** [comment]
Just had a try a few minutes ago, it works like a charm. Thanks guys!

**Comment by dflourusso** [comment]
Thanks a lot :-)

**Comment by grzjos** [comment]
Great! Works like a charm. Thank you!
